modules/axhal/src/platform/aarch64_common/psci.rs

```diff
@@ -21,22 +46,123 @@ fn psci_hvc_call(func: u32, arg0: usize, arg1: usize, arg2: usize) -> usize {
-    assert_eq!(psci_hvc_call(PSCI_CPU_ON, id, entry, arg), 0);
-    debug!("Started core {}!", id);
+///
+/// PSCI v0.1
+/// cpu entry_point: phys_addr(secondary_entry)
+pub fn psci_cpu_on(cpuid: usize, entry_point: usize, arg: usize) -> i32 {
+    info!("Starting CPU {} ON ...", cpuid);
+    let err = psci_0_1_cpu_on(PSCI_0_2_FN64_CPU_ON, cpuid, entry_point, arg);
+    if err != 0 {
+        error!("failed to boot CPU {} ({})", cpuid, err);
+    } else {
+        info!("Started CPU {}", cpuid);
+    }
+    err
+}
+
+pub fn psci_cpu_off(_cpuid: usize) {
+    const PSCI_POWER_STATE_TYPE_STANDBY: u32 = 0;
+    const PSCI_POWER_STATE_TYPE_POWER_DOWN: u32 = 1;
+    const PSCI_0_2_POWER_STATE_TYPE_SHIFT: u32 = 16;
+    let state: u32 = PSCI_POWER_STATE_TYPE_POWER_DOWN << PSCI_0_2_POWER_STATE_TYPE_SHIFT;
+    psci_0_1_cpu_off(PSCI_0_2_FN_CPU_OFF, state);
+}
+
+pub fn invoke_psci_fn(fn_id: usize, arg0: usize, arg1: usize, arg2: usize) -> usize {
+    psci_call(fn_id, arg0, arg1, arg2)
+}
+
+pub fn psci_0_1_cpu_on(fn_ids: u32, cpuid: usize, entry_point: usize, arg: usize) -> i32 {
+    let err: i32 = invoke_psci_fn(fn_ids as usize, cpuid, entry_point, arg) as i32;
+    psci_errno_tran(err)
+}
+
+pub fn psci_0_1_cpu_off(fn_ids: u32, state: u32) -> i32 {
+    let err: i32 = invoke_psci_fn(fn_ids as usize, state as usize, 0, 0) as i32;
+    psci_errno_tran(err)
+}
+
+pub fn psci_0_1_cpu_suspend(fn_ids: u32, state: u32, entry_point: usize) -> i32 {
+    let err: i32 = invoke_psci_fn(fn_ids as usize, state as usize, entry_point, 0) as i32;
+    psci_errno_tran(err)
+}
+pub fn psci_0_1_migrate(fn_ids: u32, cpuid: usize) -> i32 {
+    let err: i32 = invoke_psci_fn(fn_ids as usize, cpuid, 0, 0) as i32;
+    psci_errno_tran(err)
+}
+
+pub fn psci_errno_tran(errno: i32) -> i32 {
+    //EINVAL
+    let mut ret = -22;
+    let errno_tran = match errno {
+        0 => {
+            ret = 0;
+            PsciRetNo::PSCI_RET_SUCCESS
+        }
+        // EOPNOTSUPP
+        -1 => {
+            ret = -95;
+            PsciRetNo::PSCI_RET_NOT_SUPPORTED
+        }
+        -2 => {
+            ret = -22;
+            PsciRetNo::PSCI_RET_INVALID_PARAMS
+        }
+        // EPERM
+        -3 => {
+            ret = -1;
+            PsciRetNo::PSCI_RET_DENIED
+        }
+        -4 => PsciRetNo::PSCI_RET_ALREADY_ON,
+        -5 => PsciRetNo::PSCI_RET_ON_PENDING,
+        -6 => PsciRetNo::PSCI_RET_INTERNAL_FAILURE,
+        -7 => PsciRetNo::PSCI_RET_NOT_PRESENT,
+        -8 => PsciRetNo::PSCI_RET_DISABLED,
+        -9 => {
+            ret = -22;
+            PsciRetNo::PSCI_RET_INVALID_ADDRESS
+        }
+        _ => {
+            warn!("Unknown psci errno: {}", errno);
+            PsciRetNo::PSCI_RET_INVALID_PARAMS
+        }
+    };
+    info!("return: {:?}", errno_tran);
+    ret
+}
+
+/// PSCI return values, inclusive of all PSCI versions
+#[allow(non_camel_case_types)]
+#[derive(Debug)]
+pub enum PsciRetNo {
+    PSCI_RET_SUCCESS = 0,
```

Comment:
Returning `PSCI_RET_SUCCESS` indicates success, and it should not be a member of the error type.